content/blog/2022-04-14-self/#index.md#

```diff
@@ -0,0 +1,275 @@
+differences between object types. Normal objects have a sequence of slot
+contents for mutable slots. Object arrays then have the length, followed by the
+literal elements (pointers if objects). Byte arrays are similar, but the
+contents are padded out to 32 bit word boundaries.
+
+#### Map layout
+
+<p align="center">
+<img src="map-layout.png" />
+</p>
+
+Maps also have a slightly varying layout depending on the type. All maps share
+the same map, which is referred to as the "map map." They also need to have a GC
+link so that the GC can traverse all of the maps.
+
+
+### Bytecode
+This is relatively uninteresting, but should be mentioned. Self code is compiled
+to a bytecode that is at the same time very simple, and also expressive. It is
+almost just a syntax rewrite of the message passing with everything being on a
+stack, so this translation phase has a lot more to do with ease of interpreting
+and simplicity of the representation than some earth shattering transformation.
+
+### How it goes fast
+
+The real interesting part of this paper is how it makes Self go so
+fast. Previous work on making Smalltalk quicker was mostly focused on adding
+static types that the compiler can take advantage of, but this defeats the
+purpose of having such a dynamic language, so the authors employ a JIT with
+surprisingly few optimizations.
+
+
+#### Customized compilation
+
+The first optimization presented in this paper is "customized compilation,"
+where when a method is run with a certain type of receiver, this is noted and
+compiled for that specific type, where other message sends can be disambiguated
+at compile time. This is a huge advantage since a lot of the overhead in
+dynamically typed languages like this comes from having to constantly check the
+types of everything to determine which code to run, but in this case once you do
+something multiple times with the same types, you will have fast, specialized
+machine code to run.
+
+#### Message inlining
+
+Sometimes messages lookups can be resolved at compile time[^1], and in this case
+they are inlined. Methods are inlined as code if they are short enough and
+non-recursive. Constant data can have it's value inlined, whereas mutable reads
+and writes can be inlined as a single machine instruction (either load or
+store), since the offset is known. This is where the maps come in handy for
+giving the compiler information that is easy to work with.
+
+
+#### Message splitting
+
+A lot of the efficiency of this implementation comes from this specialized
+compilation that allows message inlining, but this nice type information is lost
+at cfg merge points, as you may have two specializations that merge together and
+then share code. The idea of message splitting is to push those merge points
+later, and then optimize the resulting code that has more type information,
+resulting in more specialized code. This may seem in general compilers not to be
```

Comment:
Maybe a brief example would help illustrate what "pushing merge points" means and how it leads to "more type information"?